### PR TITLE
Issue #1032: [UX] New text format configuration UI. 

### PR DESCRIPTION
**Pull request by docwilmot**
Fix For https://github.com/backdrop/backdrop-issues/issues/1032
Replaces https://github.com/backdrop/backdrop/pull/1012

### CODE REVIEW

**Comment by docwilmot** [comment]
@quicksketch, @jenlampton please advise if this approach seems reasonable. I don't want to work on test fixes till I'm sure I don't have to start over with a different strategy.

**Comment by docwilmot** [comment]
Screenshots on #1012

**Comment by quicksketch** [comment]
Thanks @docwilmot. Feedback posted to https://github.com/backdrop/backdrop-issues/issues/1032#issuecomment-131275916

File: core/modules/filter/filter.admin.inc [hunk 1]
```diff
@@ -152,6 +152,9 @@ function filter_admin_format_form($form, &$form_state, $format) {
   $form['#submit'] = array();
   $form['#tree'] = TRUE;
   $form['#attached']['library'][] = array('filter', 'filter.admin');
+  $form['#attached']['js'][] = backdrop_get_path('module', 'filter') . '/js/filter.admin.js';
+  $form['#attached']['css'][] = backdrop_get_path('module', 'filter') . '/css/filter.css';
+  backdrop_add_library('system', 'ui.dialog');
```
Comment (quicksketch) [inline core/modules/filter/filter.admin.inc:157]: Let's be consistent here and use:

```
$form['#attached']['library'][] = array('system', 'ui.dialog');
```

Or better, use the Backdrop dialog library to help match the existing Backdrop styling:

```
$form['#attached']['library'][] = array('system', 'backdrop.dialog');
```

File: core/modules/filter/filter.admin.inc [hunk 2]
```diff
@@ -287,35 +292,42 @@ function filter_admin_format_form($form, &$form_state, $format) {
...
+      '#suffix' => '</div>',
+    );
+    $form['filters'][$name]['filter']['title'] = array(
       '#markup' => $filter_info['title'],
     );
-    $form['filters']['order'][$name]['weight'] = array(
+    $form['filters'][$name]['filter']['details'] = array(
+      '#markup' => isset($format_tips[$name]['tip']) ? $format_tips[$name]['tip'] : '',
+    );
+    $form['filters'][$name]['filter']['configure'] = $configure_form;
```
Comment (quicksketch) [inline core/modules/filter/filter.admin.inc:329]: This might be a good place to add an AJAX button to validate the form and close the dialog.

``` php
$configure_form['actions']['#type'] = 'actions';
$configure_form['actions']['submit'] = array(
  '#type' => 'submit',
  '#value' => t('Update settings'),
  '#limit_validation_errors' => array($configure_form['#parents']),
  '#ajax' => array(
      'event' => 'click',
      'callback' => 'filter_admin_filter_form_ajax_update',
      'progress' => array('type' => 'throbber', 'message' => NULL),
  ),
  '#submit' => array('filter_admin_filter_form_submit'),
);
```

Then using `filter_admin_filter_form_submit`, have it rebuild the form. And in `filter_admin_filter_form_ajax_update`, have it pull out the parts of the form that need to be updated and then close the dialog. This is probably going to look a lot like the existing [layout_ajax_form_save_dialog](https://api.backdropcms.org/api/backdrop/core!modules!layout!layout.admin.inc/function/layout_ajax_form_save_dialog/1) function.

Unfortunately, AJAX saving and form updating is not easy. This will make the PR quite a bit more complicated.

File: core/modules/filter/js/filter.filtered_html.admin.js [hunk 3]
```diff
@@ -90,6 +90,7 @@ Backdrop.behaviors.filterFilterHtmlUpdating = {
           }
         }
         that.userTags = userTags;
+       console.log(userTags);
```
Comment (quicksketch) [inline core/modules/filter/js/filter.filtered_html.admin.js:93]: Stray `console.log` here.

File: core/modules/filter/filter.admin.inc (hunk 2 above)
Comment (docwilmot) [inline core/modules/filter/filter.admin.inc:329]: I just had a try, and youre right, I cant handle this.
I've gotten as far as opening the configuration forms with:

```
    $form['filters'][$name]['configure_link'] = array(
      '#type' => 'submit',
      '#ajax' => array(
        'callback' => 'filter_admin_ajax_form_open_dialog',
        'progress' => array('type' => 'throbber', 'message' => NULL),
```

And 

```
function filter_admin_ajax_form_open_dialog(&$form, &$form_state) {
  $filter = substr($form_state['clicked_button']['#name'], 10);
    $html = backdrop_render($form['filters'][$filter]['filter']['configure']);
    $title = "title";
    $commands[] = ajax_command_open_modal_dialog($title, $html);

  return array(
    '#type' => 'ajax',
    '#commands' => $commands,
  );
}
```

But every time you click the submit buttonin the dialog, the throbber runs and PHP logs say `PHP Notice:  Undefined index: form_build_id in C:\wamp\www\backdrop\backdrop\core\includes\ajax.inc on line 317`

The submit and callback functions dont seem to run, although I cant quite figure how to tell since dpm() is not very helpful with Ajax forms it seems.

Comment (docwilmot) [inline core/modules/filter/filter.admin.inc:329]: I'm going to roll this back to a single draggable table without the "disabled" section, and remove all the js involved. Then I'll have to leave it there, because the Ajax form stuff will influence the rest of the process (dialog, validation, final $form_state structure for saving, etc) and I really am completely at a loss with Ajax.

Sorry to bail out at this stage.

Comment (quicksketch) [inline core/modules/filter/filter.admin.inc:329]: Not a problem @docwilmot! We'll leave this one to pick up later. I'd like
to give this a shot if there's time before 1.2.0. Right now my focus is on
Link and Email fields and finishing up the WYSIWYG image handling.

On Tue, Aug 18, 2015 at 7:38 AM docwilmot notifications@github.com wrote:

> In core/modules/filter/filter.admin.inc
> https://github.com/backdrop/backdrop/pull/1016#discussion_r37305657:
> 
> > ```
> >    '#markup' => $filter_info['title'],
> >  );
> > ```
> > -    $form['filters']['order'][$name]['weight'] = array(
> > -    $form['filters'][$name]['filter']['details'] = array(
> > -      '#markup' => isset($format_tips[$name]['tip']) ? $format_tips[$name]['tip'] : '',
> > -    );
> > -    $form['filters'][$name]['filter']['configure'] = $configure_form;
> 
> I'm going to roll this back to a single draggable table without the
> "disabled" section, and remove all the js involved. Then I'll have to leave
> it there, because the Ajax form stuff will influence the rest of the
> process (dialog, validation, final $form_state structure for saving, etc)
> and I really am completely at a loss with Ajax.
> 
> Sorry to bail out at this stage.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/backdrop/backdrop/pull/1016/files#r37305657.
> 
> ## 

## 

Sent from mobile
